doc/cookbook/faas.md

```diff
@@ -0,0 +1,236 @@
+* Traffic originted: Original near traffic, easier to integrate
+* Resource saving: reusing Easegress+K8s machine resources.
+* Pay what you used: reducing small customize features' developing and maintenance cost.
+
+## Examples
+
+
+### Scenario 1: Run a FaaS function beside Easegress
+
+* After implementing your business logic
+
+1. Create a FaaSController[2]
+
+``` bash
+echo 'name: faascontroller
+kind: FaaSController
+provider: knative             # FaaS provider kind, currently we only support Knative
+syncInterval: 10s
+
+httpServer:
+    http3: false
+    port: 10083
+    keepAlive: true
+    keepAliveTimeout: 60s
+    maxConnections: 10240
+
+knative:
+   networkLayerURL: http://${knative_kourier_clusterIP}
+   hostSuffix: example.com '| egctl object create
+```
+
+2. Deploy a function into Easegress and Knative, prepare a YAML content as below:
+
+``` yaml
+name:           "demo"
+image:          "${image_url}"
+port:           8089
+autoScaleType:  "concurrency"
+autoScaleValue: "100"
+```
+
+* Save it into `/home/easegress/function.yaml`, using command to deploy it in Easegress:
+**Note** this command should be run in Easegress' instance environment and ${eg_apiport} should be replaced with the real working port, ${image_url} should be replaced with pullable image URL.
+
+``` bash
+$ curl --data-binary @/home/easegress/function.yaml -X POST -H 'Content-Type: text/vnd.yaml' http://127.0.0.1:${eg_apiport}/apis/v1/faas/faascontroller
+```
+
+3. Get the function's status, make sure it is in `active` status
+
+``` bash
+$ curl http://127.0.0.1:${eg_apiport}/apis/v1/faas/faascontroller/demo
+spec:
+  name: demo
+  image: dev.local/demo:1.0
+  port: 8089
+  autoScaleType: rps
+  autoScaleValue: "1000"
+  minReplica: 0
+  maxReplica: 0
+  limitCPU:
+  limitMemory:
+  requestCPU:
+  requestMemory:
+  requestAdaptor:
+    host: ""
+    method: ""
+    header:
+      del: {}
+      set: {}
+      add: {}
+    body: ""
+status:
+  name: demo10
+  state: active
+  event: ready
+  extData: {}
+fsm: null
+```
+
+4. Request the faas function by Easegress HTTP traffic gate with `X-FaaS-Func-Name: demo` in HTTP header.
+**Note**: this example HTTP backend's API works on `/tomcat/job/api` path and its business logic is echoing back your request body and with `V3 Body is ` content.
+
+``` bash
+$ curl http://127.0.0.1:10083/tomcat/job/api -H "X-FaaS-Func-Name: demo" -X POST -d ‘{"megaease":"Hello Easegress+Knative"}’
+V3 Body is
+‘{megaease:Hello Easegress+Knative}’%
+```
+
+### Scenario 2: Limit FaaS function resources using
+* You want to make sure at the maximum instance number can only be under 50, and it can only "180m" CPU and "100Mi" memory usage maximum allowed per instance. To providing meaningful resources amount for the function, you also want to make sure one instance has at least "100m" CPU and "50mi" memory provision. (The CPU and memory limitation usage value comes from Kubernetes resource).
```

Comment:
```suggestion
* You want to make sure at the maximum instance number can only be under 50, and it can only "180m" CPU and "100Mi" memory usage maximum allowed per instance. To providing meaningful resources amount for the function, you also want to make sure one instance has at least a "100m" CPU and "50mi" memory provision. (The CPU and memory limitation usage value comes from Kubernetes resource).
```